### ISSUE TITLE
Fake It Easy not working with .NET Core 2.0 Preview 2 with NUnit

### ISSUE THREAD

**Issue by cmaslen** [FakeItEasy/FakeItEasy#1158]
Hey,

I'm trying to get Fake it Easy to work with .NET Core 2.0 Preview 2 along with NUnit. Steps to reproduce:

- Get VS 2017 15.3.0 Preview 3.0
- Create a new .NET core 2.0 project. Note if you create a library you'll need to change it manually to a .NETCore app rather than a .NET Standard lib. See http://www.alteridem.net/2017/05/04/test-net-core-nunit-vs2017/ for details.
- In Nuget dependencies add the following:
-- Microsoft.NET.Test.SDK v15.0.0
-- NUnit v3.7.1
- Compile should at this point be ok.
- Add FakeItEasy 3.4.2

Result from the package manager log is
Restoring packages for _(project path here)_
Package restore failed. Rolling back package changes for '_(project name here)_'.

**Comment by thomaslevesque** [comment]
Hey @cmaslen,

Thanks for reporting this. We haven't looked at .NET Core 2.0 compatibility yet, I expected it to just work. We'll look into it.

**Comment by thomaslevesque** [comment]
Here's the output I see when I run `dotnet restore` with the 2.0-preview2 SDK:

```
E:\tmp\testcore2> dotnet restore
  Restoring packages for E:\tmp\testcore2\testcore2.csproj...
E:\tmp\testcore2\testcore2.csproj : error NU1605: Detected package downgrade: System.Collections.Concurrent from 4.3.0 to 4.0.12. Reference the package directly from the project to select a different version. \r
E:\tmp\testcore2\testcore2.csproj : error NU1605:  testcore2 (>= 1.0.0) -> FakeItEasy (>= 3.4.2) -> Castle.Core (>= 4.1.0) -> NETStandard.Library (>= 1.6.1) -> System.Collections.Concurrent (>= 4.3.0) \r
E:\tmp\testcore2\testcore2.csproj : error NU1605:  testcore2 (>= 1.0.0) -> FakeItEasy (>= 3.4.2) -> System.Collections.Concurrent (>= 4.0.12)
  Restore failed in 1,32 sec for E:\tmp\testcore2\testcore2.csproj.
```

I did as suggested in the error message and added a reference to `System.Collections.Concurrent` 4.3.0:

```xml
    <PackageReference Include="System.Collections.Concurrent" Version="4.3.0" />
```

With this change, restoring the packages now succeeds.

Note that the issue should already be fixed on the `develop` branch, where FakeItEasy now references `System.Collections.Concurrent` 4.3.0.

**Comment by thomaslevesque** [comment]
Interestingly, the .NET Core 1.0.4 SDK gives roughly the same message about package downgrade, but it's only a warning, not an error.

I wonder if we would have the same issue if we referenced `NETStandard.Library` instead of referencing some specific packages.

**Comment by cmaslen** [comment]
Hi @thomaslevesque ,

Thanks for the quick reply. Do you publish a Nuget package out of your develop branch? I tried searching pre-release packages but didn't see anything. I need to build my app in Docker Containers and would prefer to have no binaries in my Git repo.

**Comment by blairconrad** [comment]
@cmaslen, Appveyor does publish a feed for us. Check the README.md for the
URL. I haven't tried using it since we switched to GitFlow, but it _looks_
okay to me.

If that doesn't suit your needs, M. Levesque and I were talking about
releasing a 4.0 beta soon. If he or Mr. Ralph are available, we could do so
in the next day or so...

**Comment by blairconrad** [comment]
@thomaslevesque, the FAQ does [suggest referencing the NETStandard.Library](https://github.com/dotnet/standard/blob/c15d5dec0e9230a8a7c2abee9f7cc17c09c59804/docs/faq.md#should-i-reference-the-meta-package-or-should-i-reference-individual-packages) in stead of individual packages. Although it seems that moving forward there will be no need for this either.

I think we can consider this issue closed. If we want to reconsider how to reference platform dependencies, we can do it in a separate issue.

**Comment by blairconrad** [comment]
In case anyone's curious, the bug was fixed as a side effect of #1149.

**Comment by thomaslevesque** [comment]
> @thomaslevesque, the FAQ does suggest referencing the NETStandard.Library instead of individual packages. Although it seems that moving forward there will be no need for this either.

I find this page a little confusing... Does it mean we don't need to reference anything at all from NETStandard.Library?

> I think we can consider this issue closed. If we want to reconsider how to reference platform dependencies, we can do it in a separate issue.

👍

**Comment by blairconrad** [comment]
> I find this page a little confusing... Does it mean we don't need to reference anything at all from NETStandard.Library?

Me too. I think it means that we eventually won't. I'm not sure the current tooling will work without the references. Me, I'd just keep what we have for now. If later we see tooling that will work without the references, we could remove. But if anyone keeps using the current production tooling, I'm afraid they will need either a NETStandard.Library reference or the individual packages. Maybe?

**Comment by thomaslevesque** [comment]
> In case anyone's curious, the bug was fixed as a side effect of #1149.

In fact, it's a combination of #1190 and #1147... #1149 was just about the analyzers ;)

> Me too. I think it means that we eventually won't. I'm not sure the current tooling will work without the references. Me, I'd just keep what we have for now. If later we see tooling that will work without the references, we could remove. But if anyone keeps using the current production tooling, I'm afraid they will need either a NETStandard.Library reference or the individual packages. Maybe?

I guess we can leave it as is

**Comment by cmaslen** [comment]
Hey @thomaslevesque and @blairconrad,

Thanks again for looking into the issue. I agree leaving as it at least until the official release of Core & Standard 2.0 is the way to go. The reason I reference Core instead of Standard is for the NUnit2TestAdapter to run. This way I get my tests running from a dotnet test command and on a build server.

Very nice mocking / stubbing API by the way.

Christian.

**Comment by blairconrad** [comment]
> Very nice mocking / stubbing API by the way.

Glad you like it, @cmaslen. Of course, most of it predates our involvement with the project, but we like to think we haven't ruined it!

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.0.0).

@cmaslen, look for your name in the release notes! :trophy:
